azpubsub/contextvalidator/src/main/scala/kafka/security/auth/AzPubSubAclAuthorizer.scala

```diff
@@ -0,0 +1,459 @@
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+
+package kafka.security.auth
+
+import java.net.InetAddress
+import java.nio.charset.StandardCharsets
+import java.text.SimpleDateFormat
+import java.time.temporal.ChronoUnit
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import java.util.{Date, Locale, TimeZone}
+
+import azpubsub.kafka.security.auth.TokenValidator
+import com.yammer.metrics.core.Gauge
+import kafka.common.{NotificationHandler, ZkNodeChangeNotificationListener}
+import kafka.metrics.KafkaMetricsGroup
+import kafka.network.RequestChannel.Session
+import kafka.security.auth.AzPubSubAclAuthorizer.VersionedAcls
+import kafka.server.KafkaConfig
+import kafka.utils.CoreUtils.{inReadLock, inWriteLock}
+import kafka.utils.{CoreUtils, Json}
+import kafka.zk.{AclChangeNotificationSequenceZNode, AclChangeNotificationZNode, KafkaZkClient}
+import org.apache.kafka.common.security.auth.KafkaPrincipal
+import org.apache.kafka.common.utils.Time
+import org.slf4j.LoggerFactory
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.util.{Random, Try}
+
+object  AzPubSubAclAuthorizer{
+  val AzPubSubAclAuthorizingRequest = "AzPubSubAclAuthorizingRequest"
+  val TokenDesearializationFailRateMs = "TokenDesearializationFailRateMs"
+  val Saml2TokenTimeFormatter = "M/dd/yyyy hh:mm:ss a"
+  val TokenInvalidFromDatetimeRateMs = "TokenInvalidFromDatetimeRateMs"
+  val TokenExpiredRateMs = "TokenExpiredRateMs"
+  val TopicAuthorizationUsingTokenSuccessfulRateMs = "TopicAuthorizationUsingTokenSuccessfulRateMs"
+  val TokenNotAuthorizedForTopicRateMs = "TokenNotAuthorizedForTopicRateMs"
+  val TokenValidatorClassPathKey = "token.validator.class"
+
+  val ZkUrlProp = "authorizer.zookeeper.url"
+  val ZkConnectionTimeOutProp = "authorizer.zookeeper.connection.timeout.ms"
+  val ZkSessionTimeOutProp = "authorizer.zookeeper.session.timeout.ms"
+  val ZkMaxInFlightRequests = "authorizer.zookeeper.max.in.flight.requests"
+  val ValidateTokenInMinutes = "validate.token.in.minutes"
+
+  case class VersionedAcls(acls: Set[Acl], zkVersion: Int)
+}
+
+
+class AzPubSubAclAuthorizer extends Authorizer with KafkaMetricsGroup {
+
+  private val LOGGER = LoggerFactory.getLogger(classOf[AzPubSubAclAuthorizer])
+  private var brokerHosts  = Set[String]()
+  private var cacheTokenLastValidatedTime = new mutable.HashMap[String, Date];
+  private var tokenAuthenticator : TokenValidator = null
+  private var periodToValidateTokenInMinutes : Int = 60
+  private var zkClient: KafkaZkClient = null
+  private var aclChangeListener: ZkNodeChangeNotificationListener = null
+  private val aclCache = new scala.collection.mutable.HashMap[Resource, AzPubSubAclAuthorizer.VersionedAcls]
+  private val lock = new ReentrantReadWriteLock()
+  private val retryBackoffMs = 100
```

Comment:
Shouldn't these parameters be constants?

Reply:
retry BackOffMs? This code is completely copied from the built-in SimpleAclAuthorizer. I don't need to change anything here.